NRF52 Repeater Powersaving

fschrempf opened the pull request:
In order to reduce the power consumption of NRF52 repeaters, this implements CPU idling by suspending the main task that runs the Arduino `loop()`.

During the idle interval the CLI and the UI (if available) will be unresponsive. The RF module will be kept in RX mode and upon receiving a packet, the interrupt will cause the processing loop to continue immediately before going back to idle after all outgoing packets have been transferred.

On a RAK4631 repeater this can reduce the power consumption during RX mode from around 12 mA to around 7.5 mA.

Latest build artifacts from CI (updated 01/13/26): https://github.com/fschrempf/MeshCore/actions/runs/20971258478/artifacts/5118537314

Feedback, tests, reviews and questions welcome!

fschrempf commented:
> But it is quite hard to power down NRF52 and be waken up by RX events.

Yes, there is no real sleep mode for NRF52 so either put the CPU in idle to save power or shut it down completely. The latter requires full reinit at wakeup.

> The power consumption will be around 6mA.

In which case? With the MCU shut down and only the RF module running in RX mode? That would be in that ballpark, yes.

>  NRF52 does not handle Rising High well.

Sorry, I don't get what you mean here.

>  Your PR looks complex. May be due to this issue.

Actually it doesn't look complex to me at all. It's pretty straight forward. Instead of letting the CPU run continuously, it stops it and resumes it after the idle interval is over or a packet is received. What exactly looks complex to you?

> So I ended up just do some power saving trick to keep the power down to 8.5mA with few code changes only.

What "trick" would that be? Do you mean `waitForEvent()` [here](https://github.com/IoTThinks/MeshCore/commit/2c878edc45ac0dad6806811ac6e721c6b4c8cd33#diff-04e0d6ad57930580cf33631bfb7532c478ac1ab20b0d41ec0d319489cf207aedR100-R104)? It doesn't work for me. I'm still seeing around 14 mA with your PowerSaving07 branch.

And if it would work, it would do the same thing: put the CPU in idle, right? My approach does it in a platform-agnostic way, which is better IMHO.

IoTThinks commented:
May be I will add the CLI for esp32 based boards.

Should we share the same CLI to enable/disable power saving to both esp32 and nrf52?

Like set powersave 1?

In your PR, you set the idle period? How about wake up period?

IoTThinks commented:
@ngavars You have to measure the current at the battery cable by power meter.

A usbc, it needs to turn on unneccessary components like uart chip, led...

fschrempf commented:
> What is the intended reasonable idle interval? Something like 1 to 3 seconds?

The `idle.interval` setting is in seconds, so 10000 will give you around 2.8 hours. I'm currently running my test repeater with 1800 (30 minutes), but I'm not yet sure what would be a good value. If you have the value set very high, the automatic adverts of the repeater might be delayed accordingly (until the next wakeup happens). Apart from that I currently don't see any other downsides.

mtlynch commented on src/helpers/CommonCLI.h:
```diff
@@ -48,6 +48,7 @@ struct NodePrefs { // persisted to file
   uint8_t advert_loc_policy;
   uint32_t discovery_mod_timestamp;
   float adc_multiplier;
+  uint32_t idle_interval; // in seconds
```
Could we make this `idle_interval_seconds` and not need the comment?

fschrempf replied:
We could and I thought about it. In the end I decided to follow the style of the existing variables.

fschrempf replied:
I will leave that as is for now but happy to change it if others are in favor, too.

mtlynch commented on src/Dispatcher.h:
```diff
@@ -167,6 +167,9 @@ class Dispatcher {
   Packet* obtainNewPacket();
   void releasePacket(Packet* packet);
   void sendPacket(Packet* packet, uint8_t priority, uint32_t delay_millis=0);
+  uint32_t getTotalOutboundCount() {
```
This is exposing more data than the API client currently needs. Could it simplify to `bool hasOutboundPackets() { ... }`

That simplifies the callsite to:

```c++
if (the_mesh.getIdleInterval() &&
      ((millis() - active_timestamp) > active_time_ms) &&
      the_mesh.hasOutboundPacket()) {
    radio_driver.blockTaskUntilRXEvent(the_mesh.getIdleInterval() * 1000);
    ...
  }
```

It's easier to reason about negating a `bool` than negating a `uint32_t`.

fschrempf replied:
Good idea, yes.

fschrempf commented on examples/simple_repeater/main.cpp:
```diff
@@ -117,4 +126,11 @@ void loop() {
   rtc_clock.tick();
+
+  if (the_mesh.getIdleInterval() && (millis() - active_timestamp > active_time_ms) &&
+      !the_mesh.getTotalOutboundCount()) {
```
I will try to improve the readibilty. Though this is what clang-format suggested using the project's configuration.

fschrempf replied:
Added the additional parentheses, but left the line breaks as clang-format suggested.

ngavars commented:
I did some additional testing with power profiler and Heltec V3 board on stock firmware and on your firmware. Both measurements were captured after the initial boot/advert sequence, when the board starts idling. Heltec was powered through its battery connector with voltage set at 3307 mV. 

With stock firmware the avg current is 45 mA.
With power option firmware the avg current is also around 45 mA until approx. 3 minute mark, when it drops down to around  35 mA.

With stock firmware (latest from web flasher):
<img width="764" height="470" alt="image" src="https://github.com/user-attachments/assets/beb7f8a6-4dfb-45f7-9c27-4db7d91ffaf7" />

With power option firmware, idle.interval=1800:
<img width="764" height="470" alt="image" src="https://github.com/user-attachments/assets/61a1b4e7-30fd-40d0-88bf-aa8eaad6c63a" />

Hope this helps. I can also test it with Promicro board if you want.

4np reviewed (commented):
Thanks for working on this @fschrempf  👍  It would be good to see some more power saving for my Solar RAK19003 node :)

Having said that, I wonder if if would be better if the idle interval would be computed automatically rather than using a fixed duration. For example, in The Netherlands the mesh has grown exponentially. Where 2 months ago the mesh saw some activity, now I see hundreds of daily messages (let alone adverts) and (some parts of) Belgium and Germany have been connected. I wouldn't be surprised to see more of Germany and France starting to appear. An idle interval that worked a couple of weeks ago (maybe even days ago) may not work as well today.

Additionally, if CPU idling is triggered, how would one be able to enter remote management mode? IMHO when trying to use remote management or fetching telemetry, idling should be cancelled and the idle timer should be reset.

ngavars commented:
I did some more testing with Xiao NRF52840. Here's a comparison of stock firmware (dev branch) and power option firmware (idle.interval=1800). In both cases the board is idling for 1 minute and there are no radio events.

Stock FW
<img width="764" height="470" alt="image" src="https://github.com/user-attachments/assets/2781da5e-96bf-4479-bfa4-5377c15fc2fb" />

idle.interval=1800
<img width="764" height="470" alt="image" src="https://github.com/user-attachments/assets/5ed40126-94bd-40c5-95f1-65037254dc06" />

fschrempf commented:
> An idle interval that worked a couple of weeks ago (maybe even days ago) may not work as well today.

@4np How is the idle interval related to the mesh activity? During idle the repeater will still respond to incoming packets. Only outgoing packets will be left waiting until the timeout expires.

SaschaKt commented:
> > @4np How is the idle interval related to the mesh activity? During idle the repeater will still respond to incoming packets.  Only outgoing packets will be left waiting until the timeout expires.
> 
> @fschrempf , what I meant is an idle interval that works today, may not work tomorrow. If I read the PR, and correct me if I'm wrong, in your case you set the idle interval to 30 mins. So if there's no rx during a 30 min window, the CPU will start idling.
> 
> To illustrate the issue, this summer I didn't hear anything at all on MeshCore and there were almost no repeaters or companions. In September / October, we had a small regional mesh with several messages per hour. Today, we have [most of The Netherlands](https://map.meshcore.dev/?zoom=9&lat=52.1141&lon=5.3284) covered with the one of the most dense meshes in [Europe](https://map.meshcore.dev/?zoom=4&lat=50.2050&lon=12.5684) and I receive [several packets _per second_](https://analyzer.letsmesh.net/packets?region=netherlands).
> 
> So an idle timeout of, in your case, 30 mins that in my case would have been valid in September or summer would never trigger today.
> 
> As such IMHO it would be better to have an algorithm compute the best idle interval and have it adjust automatically on changing conditions. New repeaters pop up every single day.
> 
> ps. I think ill configured repeaters doing flood adverts at short intervals are the primary cause of the many packets per second, and putting a strain on the mesh. Reducing the number of configuration options could be beneficial as well.

Idle time is how long the CPU should be idle and then wake up for doing stuff, or it is waken before throw an rx. After a Rx/tx I have measured on my own a fast falloff of the current and on Rx a short time normal state, repeat, idle again.

4np commented:
Thank you for the clarification @fschrempf  and @SaschaKt  👍

fschrempf commented:
With https://github.com/meshcore-dev/MeshCore/pull/1266 being merged in dev, this has now been reworked to work on top of this. I will let the new version run on my RAK4630 test repeater and I will add a link here to the repeater firmware build as soon as the CI has finished for any of you who want to do another test themselves.

As a reminder, the new version requires powersaving to be turned on via CLI command `powersaving on`.

IoTThinks commented:
The PR looks great. 
I will do the test for this PR for my RAK4631 and T-Echo Lite next week.

Next week, I will push a PR to enable powersaving for ALL ESP32-based board too.
To do this, I also add 1 similar line for ESP32 in setFlag like you.
I guess the moderators will review that line in setFlag very carefully.

Hope we can have PowerSaving for both NRF52 and ESP32 repeaters soon.

4np commented:
> Here is a [link to the repeater firmware archive](https://github.com/fschrempf/MeshCore/actions/runs/20656799231/artifacts/5007854971) from the CI for any of you who want to do another test themselves.

It's running on my spare RAK based repeater and so far all seems to be working well.

fschrempf commented:
@4np Thanks for testing. On my repeater it seems like the node becomes unresponsive after some time. It might be unrelated, but I need to do some debigging.

4np commented on src/helpers/NRF52Board.cpp:
```diff
@@ -17,10 +18,20 @@ static void disconnect_callback(uint16_t conn_handle, uint8_t reason) {
   startup_reason = BD_STARTUP_NORMAL;
 }
 
+void NRF52Board::sleep(uint32_t secs) {
```
Psuedo code:

```C
static SoftwareTimer sleep_timer;
static bool sleep_timer_inited = false;
static volatile bool sleeping = false;

static void sleep_timer_callback(TimerHandle_t) {
  sleeping = false;
  resumeLoop();
}

void NRF52Board::sleep(uint32_t secs) {
  const uint32_t ms = secs * 1000UL;

  sleeping = true;

  if (!sleep_timer_inited) {
    sleep_timer.begin(ms, sleep_timer_callback, NULL, false);
    sleep_timer_inited = true;
  } else {
    // re-arm without recreating if possible:
    sleep_timer.stop();               // cancel any previous expiry
    sleep_timer.setPeriod(ms);        // if available; otherwise re-begin below
  }

  // If your SoftwareTimer wrapper does NOT have setPeriod(),
  // then do begin() again, but still guard start().
  // sleep_timer.begin(ms, sleep_timer_callback, NULL, false);

  // Critical guard: only suspend if start succeeds.
  if (!sleep_timer.start()) {
    sleeping = false;
    return; // don't brick yourself
  }

  suspendLoop();
}

void NRF52Board::wakeFromInterrupt() {
  if (!sleeping) return;
  sleeping = false;
  sleep_timer.stop();
  resumeLoop();
}
```

4np reviewed (commented):
As the radio appears to become unresponsive after a period of low RX (overnight, at least in my case) it may be that the timer wasn’t actually armed (create/start/reset failed or got wedged), but `suspendLoop()` is still getting called. With no RX there’s no other wake source and the repeaters stays suspended forever. 

The question is why it does not wake by RX. It could be that the RX interrupt isn’t firing at all?

IoTThinks commented:
I have merged a minimum piece of this code.
Yes, it can achieve 6mA.

This check "digitalRead(PIN_LORA_DIO1) == HIGH" seems to fix the "stuck" problem.
Else if we keep "ping" the RAK4631 for a while, the board will be stuck at sleep mode.

BTW, we may need to move the definition of PIN_LORA_DIO1... into platform.ini instead.
Let me monitor it for a longer period.

The suspendLoop and resumeLoop are very neat approach.

fschrempf commented:
I've updated this PR. Here is a summary of the changes compared to the previous version:

* fix `SoftwareTimer` handling (create only one instance and update it for each sleep)
  * this was likely the reason for the lockups before as I wasn't aware that `SoftwareTimer::begin()` calls `xTimerCreate()` each time
* Add generic way to prevent sleep for critical operations and use this to prevent sleep after OTA update has been activated
* Extend `MyMesh::hasPendingWork()` to take ongoing packet handling into account
* Cosmetics: Use a board-specific callback `onRXInterrupt()` in the ISR and rename the ISR

@IoTThinks @4np Thanks for your comments. With these fixes, I don't think it is necessary to explicitly check the DIO1 pin or introduce any other workarounds.

This has been running just fine on my repeater for the past two days without any noticeable issues. I will add a link for the binaries here once the CI build has finished.

fschrempf commented:
@4np One more thing: Do you use any AI tools for you reviews and code snippets? If yes, please make sure that you will mention this in the future. I want to know when I'm dealing with feedback generated by an LLM. If this is not the case, just ignore my comment. Thanks!

fschrempf commented:
@4np Thanks for testing! That's strange. On my end with the old version the problem occurred after a few hours or a day at the latest. With the new version the repeater has worked fine for four days in a row now.

I will try to do some further investigations.

fschrempf commented:
> @fschrempf your CI build does not contain the RAK4631 repeater, it [failed to build](https://github.com/fschrempf/MeshCore/actions/runs/20933125598/job/60148828798#step:4:12118) (as did the other NRF52 variants):

Oh dear! I didn't notice because the job status is green. And locally my build worked because I had modified my Arduino core manually some weeks ago when I was doing some experiments. Therefore my local `FreeRTOSConfig.h` contained `#define INCLUDE_xTaskGetHandle 1` whereas the official version doesn't have `xTaskGetHandle` and so this approach won't work.

Therefore I now reverted to the original solution (used in the very first version of this PR, sometimes the first ideas are the best ;)) using a FreeRTOS binary semaphore. This is equally good and simple and also a recommended solution in the FreeRTOS docs.

Sorry for the additional noise. I updated the [explanatory post](https://github.com/meshcore-dev/MeshCore/pull/1238#issuecomment-3740222763) above accordingly.

Here are the latest binaries from the CI build: n. a.

fschrempf commented:
@4np Ok, but don't spend too much efforts in testing. Currently it looks like we will go for @IoTThinks alternative solution in https://github.com/meshcore-dev/MeshCore/pull/1353.

SaschaKt commented:
> Maybe I should test #1353 instead then 😄

From 8mA to 4mA, very good, I'm using it from the first version primary with v3/v4 (48mA down to 9-10mA) as solar repeater
